prov/efa/src/efa_mr.c

```diff
@@ -239,12 +239,18 @@ static int efa_mr_hmem_setup(struct efa_mr *efa_mr,
 				return err;
 			}
 		}
-		err = cuda_dev_register((struct fi_mr_attr *)attr, &efa_mr->peer.device.cuda);
-		if (err) {
-			EFA_WARN(FI_LOG_MR,
-				 "Unable to register handle for GPU memory. err: %d buf: %p len: %zu\n",
-				 err, attr->mr_iov->iov_base, attr->mr_iov->iov_len);
-			return err;
+
+		/* When cudaMemcpy is allowed (cuda_xfer_setting==CUDA_XFER_ENABLED), use cudaMemcpy (which does not need registration)
+		 * Otherwise, use gdrcopy and register the memory with gdrcopy.
+		 */
+		if (cuda_get_xfer_setting()!=CUDA_XFER_ENABLED) {
```

Comment:
What happens when someone sets `CUDA_XFER_ENABLED == TRUE` and `HMEM_CUDA_USE_GDR_COPY == TRUE`

Reply:
What happens is that we register as a cuda buffer, but the util code ries to go down the GDR copy path and we hit an assert.

Reply:
>  What happens when someone sets CUDA_XFER_ENABLED == TRUE and HMEM_CUDA_USE_GDR_COPY == TRUE

We use cudaMemcpy in that case.

> What happens is that we register as a cuda buffer, but the util code ries to go down the GDR copy path and we hit an assert.

Not sure. where is the assert?

Reply:
sorry.. im wrong